packages/service-library/src/servicelib/redis/_decorators.py

```diff
@@ -1,56 +1,148 @@
+        lock_value -- some additional data that can be retrieved by another client if None,
+                    it will be automatically filled with the current time and the client name
 
     Raises:
         - ValueError if used incorrectly
         - CouldNotAcquireLockError if the lock could not be acquired
+        - LockLostError if the lock was lost (e.g. due to Redis restart, or TTL was not extended in time)
     """
 
     if not lock_key:
         msg = "lock_key cannot be empty string!"
         raise ValueError(msg)
 
-    def decorator(func: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:
+    def _decorator(
+        func: Callable[P, Coroutine[Any, Any, R]],
+    ) -> Callable[P, Coroutine[Any, Any, R]]:
         @functools.wraps(func)
-        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+        async def _wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
             redis_lock_key = (
                 lock_key(*args, **kwargs) if callable(lock_key) else lock_key
             )
             assert isinstance(redis_lock_key, str)  # nosec
 
-            redis_client = redis(*args, **kwargs) if callable(redis) else redis
-            assert isinstance(redis_client, RedisClientSDK)  # nosec
+            client = (
+                redis_client(*args, **kwargs)
+                if callable(redis_client)
+                else redis_client
+            )
+            assert isinstance(client, RedisClientSDK)  # nosec
+            nonlocal lock_value
+            if lock_value is None:
+                lock_value = f"locked since {arrow.utcnow().format()} by {client.client_name} on {socket.gethostname()}"
 
-            async with redis_client.lock_context(
-                lock_key=redis_lock_key, lock_value=lock_value
+            lock = client.create_lock(redis_lock_key, ttl=DEFAULT_LOCK_TTL)
+            if not await lock.acquire(
+                token=lock_value,
+                blocking=blocking,
+                blocking_timeout=(
+                    blocking_timeout.total_seconds() if blocking_timeout else None
+                ),
             ):
-                return await func(*args, **kwargs)
+                raise CouldNotAcquireLockError(lock=lock)
+
+            try:
+                async with asyncio.TaskGroup() as tg:
+                    started_event = asyncio.Event()
+                    # first create a task that will auto-extend the lock
+                    auto_extend_lock_task = tg.create_task(
+                        _periodic_auto_extender(lock, started_event),
+                        name=_EXCLUSIVE_AUTO_EXTEND_TASK_NAME.format(
+                            redis_lock_key=redis_lock_key
+                        ),
+                    )
+                    # NOTE: In case the work thread is raising right away,
+                    # this ensures the extend task ran once and ensure cancellation works
+                    await started_event.wait()
```

Comment:
so this will wait until `started_event.set()` right? so if the `auto_extend_lock_task` raises, it is captured in `tg: TasGroup`?

Reply:
any exception in auto_extend_lock_task will be catched by TaskGroup. wait or no wait.
No I added this in order to be sure that my test with a work_task that raises straight away works.